### PR TITLE
SUS-521: Refactor WikiaExternalImageList extension

### PR DESCRIPTION
**Pull request by mszabo-wikia**
The extension now checks URLs against the contents of `MediaWiki:Global external image whitelist` on Community Central, whose contents it interprets the same way core MW treats MediaWiki:External image whitelist.

Ticket & QA notes: https://wikia-inc.atlassian.net/browse/SUS-521
@mixth-sense

### CODE REVIEW

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.i18n.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+$messages = [];
```
Comment (sqreek) [inline extensions/wikia/WikiaWhiteList/WikiaExternalImageList.i18n.php]: Please add `qqq` section

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	 *
+	 * @param string $url
+	 * @return bool Whether the URL matches an item in the global external image whitelist
+	 */
+	public static function onOutputMakeExternalImage( &$url ) {
+		$listPage = GlobalTitle::newFromTextCached( static::GLOBAL_WHITELIST_PAGE, NS_MEDIAWIKI, static::COMMUNITY_WIKI_ID );
+		$list = explode( "\n", $listPage->getContent() ?? '' );
+
+		// Handle whitelist items in the same way as MediaWiki:External image whitelist
+		foreach ( $list as $entry ) {
```
Comment (sqreek) [inline extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php]: Can we join entries and use `|` in regex?
Then we can use one preg_match against entry instead loop

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php [hunk 3]
```diff
@@ -28,16 +28,19 @@ public static function onOutputMakeExternalImage( &$url ) {
 		$list = explode( "\n", $listPage->getContent() ?? '' );
 
 		// Handle whitelist items in the same way as MediaWiki:External image whitelist
+		// Remove blank entries/comments
+		$i = 0;
 		foreach ( $list as $entry ) {
```
Comment (sqreek) [inline extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php]: You can use http://php.net/manual/pl/function.array-filter.php

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php [hunk 4]
```diff
@@ -28,16 +28,19 @@ public static function onOutputMakeExternalImage( &$url ) {
...
-			if ( preg_match( '/' . str_replace( '/', '\\/', $entry ) . '/i', $url ) ) {
-				return true;
+				unset( $list[$i] );
 			}
+			$i++;
 		}
 
-		return false;
+		// Combine and evaluate the list as a single regular expression
+		$combinedRegex = implode( '|', $list );
```
Comment (sqreek) [inline extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php:33]: make sure there is no `|` in $list, and add some tests for this one.

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+			return ( strpos( $entry, '#' ) !== 0 && $entry !== '' );
+		} );
+
+		// Combine and evaluate the list as a single regular expression
+		$combinedRegex = implode( '|', $list );
+
+		// Escape slashes
+		$combinedRegex = str_replace( '/', '\\/', $combinedRegex );
+
+		return preg_match( '/(' . $combinedRegex . ')/i', $url ) === 1;
```
Comment (nandy-andy) [inline extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php:38]: Are you sure you want to return in a hook something else than `true`? Returning `false` will break the hooks' chain and may generate some hidden issues but let me know if you want to discuss it offline and describe me better the big picture :)

File: extensions/wikia/WikiaWhiteList/tests/WikiaExternalImageListTest.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+class WikiaExternalImageListTest extends WikiaBaseTest {
```
Comment (nandy-andy) [inline extensions/wikia/WikiaWhiteList/tests/WikiaExternalImageListTest.php:3]: :gem: for unit tests! :)

**Comment by nandy-andy** [comment]
LGTM :+1:

File: extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php (hunk 5 above)
Comment (nandy-andy) [reply extensions/wikia/WikiaWhiteList/WikiaExternalImageList.class.php:38]: We discussed it offline:
- the hook is used in a condition in Parser,
- there is no other hooks registered.

**Comment by ghost** [comment]
This is waiting for SUS to take another round with ImageReview.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 10 minor
* ![INFO][INFO] 3 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [GlobalTitle.php#L18](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L18): Class "GlobalTitle" has 35 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
1. ![MAJOR][MAJOR] [GlobalTitle.php#L69](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L69): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. ![MAJOR][MAJOR] [GlobalTitle.php#L97](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L97): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. ![MAJOR][MAJOR] [GlobalTitle.php#L122](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L122): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. ![MAJOR][MAJOR] [GlobalTitle.php#L676](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L676): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. ![MAJOR][MAJOR] [GlobalTitle.php#L768](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L768): Remove the use of this variable variable "$$key". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1599)
1. ![MINOR][MINOR] [GlobalTitle.php#L101](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L101): Immediately return this expression instead of assigning it to the temporary variable "$title". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. ![MINOR][MINOR] [GlobalTitle.php#L318](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L318): Immediately return this expression instead of assigning it to the temporary variable "$url". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. ![MINOR][MINOR] [GlobalTitle.php#L444](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L444): Immediately return this expression instead of assigning it to the temporary variable "$text". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1488)
1. ![MINOR][MINOR] [GlobalTitle.php#L478](https://github.com/Wikia/app/blob/8d4cd53df302c5ed1244635366dd53da9c09bcb2/includes/wikia/GlobalTitle.php#L478): Remove the literal "false" boolean value. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1125)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request closed without merging**